### DISCUSSION TITLE
Reading on Analog Input/Output

### DISCUSSION THREAD

**Discussion by Asclepiios** [gijzelaerr/python-snap7#363] [Q&A] [answered]
Hello, 

For a project I need to read the Input and Output of a PLC with python. 
I do not have acces to the PLC, but my reading of Input and Output seems to work, since I obtain value and no error.
Here is my code :  
`data = s7_client.read_area(snap7.types.Areas.PE, 0, starting_byte, size)` 
Is there a way to read Analog Input/Output ? It seems not possible with snap7.types.Areas.PE since it give me the error :
`b'CPU : Address out of range' `
Is there a specific Areas that I miss ? I don't see one in the file, or any function ready to read on analog. 

with kind regards,
Asclepiios

**Comment by swamper123** [comment]
Hello there,

first question is: What device and what version of python-snap7 do you got?

**Reply by Asclepiios** [reply]
Thanks for the answer, I work on a SIEMENS PLC, a S7, CPU315-2, but the aim of the program is to work on other SIEMENS CPU as well. 
The python-snap7 version is 1.1, (and I'm working on Linux, Ubuntu 22.04 if it is of any help)

**Reply by swamper123** [reply]
Okay, that helps a bit more.  :)

PE and PA should be fine for analog in/outputs as well.
The Message you reveive comes from the SPS, so something isn't fine with the request.
Do you know how many Bytes you can read (so what is your `starting_byte` and `size` value? 
How does your PE Area looks like on PLC side?
Have you tried sth. like `s7_client.read_area(snap7.types.Areas.PE, 0, 0, 1)` which should read one byte at least?

**Reply by Asclepiios** [reply]
My PE area look like this : 
![Capture d’écran du 2022-05-24 12-18-47](https://user-images.githubusercontent.com/72105059/170009549-12488139-71be-44bd-b3fa-1576bbe54592.png)
When I read with size = 1 & starting_byte = 17 for example, I read 15 (0x0F)
With starting_byte = 0 I read 255 (0xFF)
So it seems to work, but when I read with a starting_byte = 256 & size = 2 (First analog input) I got : 
`snap7.exceptions.Snap7Exception: b'CPU : Address out of range'`

Maybe I got the starting Byte wrong ? It is the first thing I think of, but I do not see why.

**Reply by swamper123** [reply]
I haven't got a decive to test things. :/
Stupid question: Is that analog Input 12 Bit or one word long? Not sure what Kind of Datatype S7 uses ... it is a part I have never used so much before.

Perhaps @maxitaxi13 can help out, since he used PA/PE as well ...

**Reply by Asclepiios** [reply]
The analog input seems to be one word long. 
![image](https://user-images.githubusercontent.com/72105059/170023098-d093006e-355a-4eb3-925d-3564647081c6.png)
When I try to read on an address that is more than 127, I got the `b'CPU : Address out of range'` error. 
But the value between my last Input and 127 are all at 0. 

Edit :
Based on this post : https://sourceforge.net/p/snap7/discussion/general/thread/b549eeb3/?limit=25
I tried to read with the Area at 0x80, and I'm able to read at 128 and more. 
I have no idea why ? But I still think it's relevant

**Reply by swamper123** [reply]
Hmm ... no clue ... sorry I have to pass that. :confused:

**Reply by Asclepiios** [reply]
Thanks for the help anyway :) 
I will continue to see if there is something to do with that !

**Reply by Asclepiios** [reply]
Hello, 
I may have found part of my problem, so I will give an update :
PEWs are not simply analog input/output in my case but **Peripheral** Input/Output Word.
Every cycle, the Process image reads all the input and saves them. But it has a limited size, in my case 128 Byte for input, and 128 Byte for the output.

If you want to access I/O outside of this range, you must use Peripheral access. It also gives the real value a the moment, since it doesn't use Process Image.
The input/output area in snap7 may use Process Image, making it impossible to read outside of this range.
There may be another area for that, not implemented, as I was saying last time maybe the 0x80. I will try things if I happen to get a PLC in my hand. 
(The Area 0x80 is encouraged by the WireShark github, line 333 [here](https://github.com/wireshark/wireshark/blob/master/epan/dissectors/packet-s7comm.c))

(More info about Peripheral Access, and Process Image [here](https://support.industry.siemens.com/cs/document/18325417/where-and-when-do-you-need-peripheral-addressing-?dti=0&lc=en-WW))

**Comment by gijzelaerr** [comment, marked as answer]
Closing as stale. This was a general usage question about address ranges, answered in 2022.
